Add a "pre-verify" callback to ssl_tls.c

NWilson opened the pull request:
This enables a client to populate the trust chain on-demand, rather than loading all the trusted certificates up-front.  This is useful on mobile clients where the OS cert store contains >200 certificates, 199 of which won't be used at any given time.

We're using this in our clients, but I'd rather not keep any local patches. I don't think it's possible to use the existing verify callback for this, the basic issue is that by the time the existing verify callback is invoked, the chain has already been built, so it's a real hack to try and look up certs from the OS store at that point.

> Aside — in fact for a while we were using the existing verify callback to load certs from the OS store. We'd let the chain building fail, then from the verify callback we'd search the OS store, and clear the `MBEDTLS_X509_BADCERT_NOT_TRUSTED` flag if we found a root. However, for cert chains with cross-signed roots (like www.google.com), it's just not possible to get `
> mbedtls_x509_crt_verify` to build the correct chain unless the correct root certs have been fetched already, it really is too late.

Questions:
1. Are you interested in this?
2. For binary compatibility, I guess I'd need to hide this behind an ifdef, since it changes the size of `struct mbedtls_ssl_config`. Or, you could merge this into the development branch for the next major release.
3. What tests does this need - none really, since the code does nothing, but I guess a smoke test to check that the callback is invoked would be nice?

Thanks as always for your help.

sbutcher-arm commented:
Hi @NWilson,

We're interested! I think this is a bit specialised as a use case, and so it would be best to have guards around the feature, defaulting to off, which would also retain binary compatibility.

I think the tests can be simple. Just check the callback works as required, and the context is passed correctly. I don't see a negative you can test here, as obviously if the callback is unset, nothing will be called.

NWilson commented:
I've added a compile-time guard `MBEDTLS_SSL_PREVERIFY_CB`, at the end of the list in config.h (which seems not be in alphabetical order, so I hope that's right).

I've also added a test to `test_suite_ssl.function`. There don't seem to be ably high-level integration tests for ssl_tls.c, neither using the "self-test" mechanism nor in `tests/test_suite_ssl.function`, so I hope that is the correct place for this.

NWilson commented:
Ping! I did add the tests as requested.

sbutcher-arm commented:
@gilles-peskine-arm - Could you please do a design review?

sbutcher-arm commented on include/mbedtls/config.h:
```diff
@@ -1370,6 +1370,15 @@
+ *
+ * Uncomment this to enable support for the preverification callback
+ */
+//#define MBEDTLS_SSL_PREVERIFY_CB
```
We have far too many features already defined in `config.h`, and I don't think this feature is substantial enough to merit yet another config option of `MBEDTLS_SSL_PREVERIFY_CB`. I'm aware this contradicts what I said earlier. This isn't going into a LTS release, so we can afford to extend the context and break the ABI.

As written the feature is optional because if it's not set, you don't have to call these callbacks. In terms of code size, X.509 is so large, that I don't think a few extra bytes for this callback is worth worrying about. So I think I'd just extend the API with this feature, and not introduce the config option.

sbutcher-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -1,6 +1,34 @@
+}
+static unsigned char pre_vrfy_buffer[2048];
+static int pre_vrfy_buffer_start = 0, pre_vrfy_buffer_end = 0;
+static int pre_vrfy_send( void *ctx, const unsigned char *buf, size_t len )
```
I know it's test code, but we need some comments here to explain the code, at least a little.

sbutcher-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -52,4 +80,45 @@ void ssl_set_hostname_twice( char *hostname0, char *hostname1 )
\ No newline at end of file
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_SSL_PREVERIFY_CB:MBEDTLS_FS_IO:MBEDTLS_KEY_EXCHANGE_RSA_ENABLED:MBEDTLS_AES_C:MBEDTLS_SHA256_C:MBEDTLS_CIPHER_MODE_CBC */
```
Lots of dependencies! Will this test case execute in every build we need it to?

I think we may need to simplify the test. I don't want it to be unnecessarily skipped if one of the test dependencies is missing.

sbutcher-arm commented on include/mbedtls/ssl.h:
```diff
@@ -1063,6 +1069,23 @@ void mbedtls_ssl_conf_verify( mbedtls_ssl_config *conf,
+ * \param p_pre_vrfy pre-verification parameter
+ */
+void mbedtls_ssl_conf_pre_verify(mbedtls_ssl_config *conf,
+                                 void(*f_pre_vrfy)(void *, mbedtls_x509_crt *),
```
We should `typedef` this callback, and also have a doxygen definition. At the time this PR was first written we didn't do that, but we do now.

sbutcher-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -52,4 +80,45 @@ void ssl_set_hostname_twice( char *hostname0, char *hostname1 )
+    /* Read in the certificate record, and check it calls the pre-verify callback */
+    conf.endpoint = MBEDTLS_SSL_IS_CLIENT;
+    mbedtls_ssl_conf_pre_verify( &conf, pre_vrfy_fn, (void*)&pre_vrfy_data );
+    TEST_ASSERT( mbedtls_ssl_parse_certificate( &ssl ) == 0 );
```
Should test with the function set and when not set.

gilles-peskine-arm commented:
(This is a design review, I haven't reviewed the code yet.)

I don't think this needs a compile-time option. The new feature does not change the behavior and does not increase the code size by a significant amount. This is acceptable as an API change. It break the ABI but this is acceptable for a new release. We have too many compile-time options as it is, so this should not introduce a new compile-time option.

This feature has great power and so must be used responsibly. I can foresee many things that people will _want_ to do with this feature. Will all of them work? The documentation of `mbedtls_ssl_conf_pre_verify` should be expanded to make it clear what works and what doesn't, and what information can and cannot be trusted at this stage.

A real-world example: my device boots with its clock blinking 12. It makes a TLS connection which would normally fail because the certificate is not within its validity period. Can I change the system time to make the certification verification pass? Assuming it works, is it secure or have I just opened a security hole?

Patater commented:
Reviewed and approved for design.

gilles-peskine-arm commented:
There are merge conflicts. While they're easy to resolve, I'd prefer to review the code as it's going to go. @sbutcher-arm Would you be ok with rebasing this to 2.8?

sbutcher-arm commented:
@gilles-peskine-arm - That's fine. I'll rebase it, and resubmit it as a new PR.

sbutcher-arm commented:
@gilles-peskine-arm - New PR created as #1521.

cc: @AndrzejKurek
